op-deployer/pkg/deployer/verify/constructors.go

```diff
@@ -0,0 +1,53 @@
+package verify
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var constructorArgSlots = map[string]int{
+	"ProxyAdminAddress":              1,
+	"OpcmAddress":                    28,
+	"DelayedWETHImplAddress":         1,
+	"OptimismPortalImplAddress":      2,
+	"PreimageOracleSingletonAddress": 2,
+	"MipsSingletonAddress":           1,
+	"SuperchainConfigProxyAddress":   1,
+	"PermissionedDisputeGameAddress": 12,
+	"FaultDisputeGameAddress":        10,
+}
+
+func (v *Verifier) getConstructorArgs(ctx context.Context, address common.Address, contractName string) (string, error) {
```

Comment:
I'd love to find a way to unit test this function so that we can catch issues with the verification logic before changes land on develop. I think we can build a test that uses a Solidity script to encode the constructor args for each of these contracts, then compares it to what would be returned from this method.

Reply:
I was able to use @tynes suggestion to parse the abi to automatically extract the number of constructor args from there. So we no longer need the brittle hardcoded map of `contract-name --> numConstructorArgSlots`. I added a unit test for the abi parsing [here](https://github.com/ethereum-optimism/optimism/blob/9e5001f2ffca647279f4e888d7d579a3c77d7c37/op-deployer/pkg/deployer/verify/constructors_test.go). Let me know if you still think we need to add a new solidity script as part of testing or if you think the unit test I linked will suffice. Not sure if the solidity script unit test you had in mind would allow us to check the actual value of encoded args, or if you're intent there was to check the number of arg slots, which we are already checking now in that simpler test

Reply:
@mslipper I am trying to follow your comment about using a solidity script. Are you thinking that we would modify one of the existing solidity scripts (e.g. DeployImplementations.s.sol) such that its output includes the `encodedConstructorArgs` + `deploymentTxHash` of each contract it deploys? Then our go code could consume that information, and check it against what the `v.getConstructorArgs` function returns?

Reply:
Since we no longer have the brittle hardcoding the creates a unique case for each contract, we are going to lean on other unit tests for getConstructorArg helper functions to verify this functionality instead of a solidity test script